sycl/source/detail/device_image_impl.hpp

```diff
@@ -264,27 +264,31 @@ class device_image_impl {
                        NumElements ==
                    0 &&
                "unexpected layout of composite spec const descriptors");
-        auto *It = reinterpret_cast<const std::uint32_t *>(&Descriptors[8]);
-        auto *End = reinterpret_cast<const std::uint32_t *>(&Descriptors[0] +
-                                                            Descriptors.size());
+        const uint8_t *It = &Descriptors[8];
```

Comment:
DeviceBinaryProperty has a `asUint32` method.  Couldn't we call that above, instead of `.asByteArray()` and then avoid the reinterpret cast on that front?  

Alternately, instead of inserting std::memcpy down below, I wonder if we could use `sycl::bit_cast` ?  

I know the reinterpret cast for pointers means we potentially have UB, but given that we are getting raw binary data from the PI, are we really avoiding any issues by switching from reinterpet_cast to memcpy into local vars?

Reply:
> Alternately, instead of inserting `std::memcpy` down below, I wonder if we could use `sycl::bit_cast` ?

I believe `sycl::bit_cast` asserts that `To` and `From` have the same size, so we can't use it here sadly.

> I know the reinterpret cast for pointers means we potentially have UB, but given that we are getting raw binary data from the PI, are we really avoiding any issues by switching from reinterpet_cast to memcpy into local vars?

Currently there isn't must difference between the two, except we assume that the data sycl-post-link generates has the same structure as the types we use. If this starts failing it is on us. However, relying on UB means we are at the compiler's mercy and any day it could change. Do I think it's likely to happen? No, but it would likely be a huge headache if it ever does and making this change seems like a small price to avoid future pains. Better safe than sorry!